docs: new feature CYPHER preambles (#1022)
neo4j-contrib/neo4j-spark-connector: docs/modules/ROOT/pages/performance/cypher-tuning.adoc

= Cypher query tuning
label:new[New in 6.0]

> :description: This chapter covers how users can pass Cypher preambles to fine-tune performance in niche scenarios.

Reviewer: ```suggestion
[role=label--new-6.0]
= Cypher query tuning
:description: This chapter covers how users can pass Cypher preambles to fine-tune performance in niche scenarios.
```

Default Cypher parameters are optimized for the general use case.
However if you need to access advanced tuning options, this is fully supported with our connector.

> For the most up to date guide on how to use these parameters, please refer to the official Cypher manual about https://neo4j.com/docs/cypher-manual/current/planning-and-tuning/query-tuning[query tuning].

Reviewer: ```suggestion
However, the Spark connector supports all the query tuning options described in the link:https://neo4j.com/docs/cypher-manual/current/planning-and-tuning/query-tuning[Cypher manual].
```

include::partial$sparksession.adoc[]

[role=label--new-6.0]

> == Specify Cypher version

Reviewer: ```suggestion
== Specify the Cypher version
```

> If you want to read/write with a specific Cypher version, it can be done by passing a `cypher.version` option.

Reviewer: ```suggestion
You can change the Cypher version using the `cypher.version` option.
```

[.tabbed-example]
====
[.include-with-Scala]
=====
.Example
[source, scala, role=nocollapse]
----
include::example$scala/BasicCypherPreambles.scala[tags=code-cypher-version]
----
=====

[.include-with-Python]
=====
.Example
[source, python, role=nocollapse]
----
include::example$python/basic_cypher_preambles.py[tags=code-cypher-version]
----
=====
====

.Equivalent Cypher query
[%collapsible%]
====
[source, cypher]
----
CYPHER 25
MATCH (o: Object) RETURN o
----
====

[role=label--new-6.0]

> == Pass query tuning parameters

Reviewer: ```suggestion
== Add query tuning parameters
```

We offer full flexibility on parameter passing to the Cypher runtime.
Any tuning option that you can see from the Cypher manual you can pass simply by setting your Spark options using the `cypher.tuning` prefix.

> What you pass as these options are directly sent as CYPHER preambles.

Reviewer: ```suggestion
You can add any tuning option described in the Cypher manual using the `cypher.tuning` prefix in a Spark option.
All the tuning options are added to the Cypher query as `CYPHER` preambles.
```

[.include-with-Scala]
=====
.Example
[source, scala, role=nocollapse]
----
include::example$scala/BasicCypherPreambles.scala[tags=code-runtime-parallel]
----
=====

[.include-with-Python]
=====
.Example
[source, python, role=nocollapse]
----
include::example$python/basic_cypher_preambles.py[tags=code-runtime-parallel]
----
=====
====

.Equivalent Cypher query
[%collapsible%]
====
[source, cypher]
----
CYPHER runtime=parallel
MATCH (o: Object) RETURN o
----
====

You are also free to add multiple parameters this way, simply add them separately like so.

[.tabbed-example]
====
[.include-with-Scala]
=====
.Example
[source, scala, role=nocollapse]
----
include::example$scala/BasicCypherPreambles.scala[tags=code-multiple-tuning]
----
=====

[.include-with-Python]
=====
.Example
[source, python, role=nocollapse]
----
include::example$python/basic_cypher_preambles.py[tags=code-multiple-tuning]
----
=====
====

.Equivalent Cypher query
[%collapsible%]
====
[source, cypher]
----
CYPHER runtime=parallel operatorEngine=interpreted
MATCH (o: Object) RETURN o
----
====

== Tuning feature compatibility

In the above examples we illustrated using custom read query for the sake of simplicity.

> However, tuning preambles and cypher version can be used for writing as well as other operations.

Reviewer: ```suggestion
While the examples above show how to tune preambles and change the Cypher version for read queries, the same features can be used for writing as well as other operations.
```

> However, tuning preambles and cypher version can be used for writing as well as other operations.

Reviewer: ```suggestion
While the examples above show how to tune preambles and change the Cypher version for read queries, the same features can be used for writing as well as other operations.
```
